### DISCUSSION TITLE
Welcome to µCNC Discussions!

### DISCUSSION THREAD

**Discussion by Paciente8159** [Paciente8159/uCNC#228] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by rehanzfr-zz** [comment]
I am using uCNC with the mega. Everything is fine. Please let me know whether we can increase the speed of the machine in curves or circles. It is now as given below, however, still it is very slow. Should I increase these values?
$11 = 3.000    (Junction deviation, millimeters)
$12 = 1.000    (Arc tolerance, millimeters)

**Reply by Paciente8159** [reply]
Ok I'll try to explain those parameters as best as I can. They mostly work the same way (with slight differences of course) across other firmwares like Grbl vs Marlin.

[$11-juntion deviation](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#11---junction-deviation-mm)
![$11](https://user-images.githubusercontent.com/12116536/181917367-c5fb6925-1cbd-4697-8189-4a79cd2cad3c.jpg)
As stated in the wiki this modify deacceleration/acceleration calculation between two line segments.
For angles lower then 90º a speed change is applied to compensate for the direction change and minimize the effects of speed changes on the motors moving a mass. The amount of speed change depends on this angle. Closer to 0 the less deacceleration/acceleration is applied. The closer to 90º the more this deacceleration/acceleration is enforced.
For angles equal or higher then 90º at least one of the motors either stops completely or changes direction. This feels the full effect of the deacceleration/acceleration that is going from the current speed to 0 and then ramp up again.

You can modify this behavior with this. parameter. Only values between -1 and 1 will have effect (beyond that they are ignored). Higher values will make it deaccelerate/accelerate less. Lower values will enforce deaccelerate/accelerate more heavily even at lower angles.

A higher value will (towards 1) will help to go through corners (and arcs) faster but will also induce more stress to the steppers due to the abrupt speed changes.  


[$12--arc tolerance](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#12--arc-tolerance-mm)
![$12](https://user-images.githubusercontent.com/12116536/181917931-721bb649-3461-4410-bba2-95a4e8a8c8a0.jpg)
Arcs are done by multiple line segments that approximate the arc curve. The smaller the line segments the more perfect is the arc but will increase the number of arcs and iterations to generate and arc (approximation) trajectory.
This parameter is the distance between the middle of a line segment used to simplify the arc shape and the real arc curvature.

Increasing this value will reduce the number of arcs and calculations but will make your arc look more like polygons. This depends a bit on what your resolution/precision needs is. There are trade-offs on both ends for reducing or increasing this value.

If your machine is doing arcs too slowly you should try a few things before changing the default values of these settings. These are:

- Increase motors acceleration ($120,$121 and $122)
- Increase motors top speed ($110,$111 and $112)
- Change the path mode to G61.1 before starting an arc and then resetting to G61 after the arc

**Comment by topcatai** [comment]
Hello, Which post processor is recommended that we use for this?
Will this work with Universal GCode Sender (https://winder.github.io/ugs_website/)

**Reply by Paciente8159** [reply]
Hi there,

µCNC is based on Grbl protocol. So it should work with and post processor that emits Grbl compatible gcode or with any Grbl compatible Gcode sender. So it should work with UGS just fine. If you find any issue with a particular program please report.

**Reply by topcatai** [reply]
Hi, 
  Thank you for your reply. 

  One more noob question, Are you aware of any GRBL post processors that can output 4th Axis code? We are currently using LinuxCNC 
 and had to edit the postprocessor to include 4th Axis Rotary and specify its alignment against X / Y Axis.

 We are a startup, manufacturing 3&4 Axis CNC's and we are exploring alternatives to LinuxCNC for affordable controller options. We would love to support your effort as well and are keen on getting a working USB/Arduino/UGS based system.
 
 We would like to explore supporting you based on a licensing option of USD 20 per License - in other words we will contribute this amount everytime we sell a machine with this configuration in perpetuity.
 
 Kind Regards,
Tapan

https://akriti.tech

**Reply by Paciente8159** [reply]
That's awesome. Thanks..
For 4D? Apart from highend software like Solidworks, or BobCAD and other commercial products, no...I don't know any.
You might want to explore [FreeCAD post processor customizations](https://wiki.freecad.org/Path_Postprocessor_Customization).
There seems to be some sort of [4D development for FreeCAD](https://wiki.freecad.org/Path_fourth_axis) but it seems to be experimental.

**Reply by topcatai** [reply]
Hi,

Thanks for the reply. We do support all highend softwares like Fusion, Solidworks etc.
When we initially released our 4 Axis machine based on a Linux CNC board, we had to enable the 4th axis within the Linux CNC post processor file (which is present in all CAM softwares like the ones mentioned above)
We were wondering if someone on your end knows how we could pull that off with grbl.

Having said that, we will explore how this can be done. We thank you for your quick responses and hope to get this working within the next 3 months. We will keep you updated along the way and we are happy to share the full implementation with everyone here, in case we manage to pull it off.

Regards

**Reply by Paciente8159** [reply]
> We were wondering if someone on your end knows how we could pull that off with grbl.

Unfortunately no.
But there are people out there that [already done it for Fusion360](https://forums.autodesk.com/t5/fusion-360-manufacture/grbl-multi-axis-post-processor/td-p/9413041) and [here](https://github.com/dguerizec/fusion360-grbl-4axis), so I assume there are probably some implementations out there for other software as well.

**Comment by topcatai** [comment]
***@***.***

**Comment by aquietude** [comment]
I want to be able to control solenoids over G-code with uCNC based uppon Marlin M380/M381 G-codes, similar to what you did with M17/M18. What I mean is, I want to write an additional module for the uCNC which can be used similar to the module you wrote for M17/M18 G-codes, but, reading your parser_m17_m18.c code I saw you used io_enable_steppers, and an equivalent for solenoids doesn't exists under the uCNC core, and reading the Marlin firmware source code and your parser codes I didn't understood clearly how you refactored their code for working on uCNC. Could you give some guidance on how could I do so? Would that means that I need to implement an io_enable_solenoids fuction equivalent to the io_enable_steppers but for solenoids? Also, I didn't understood clearly what you based yourself for writing io_enable_steppers. I previously thank you for your attention, answer and guidance/advices regarding that.

**Reply by Paciente8159** [reply]
Hi there.

If I understood correctly you want to be able to enable and disable some solenoids.
I'm guessing these solenoids will be controlled with a simple digital IO port.

µCNC 4 types of output ports

- Special outputs - these are just normal digital output ports that have dedicated functions inside the firmware core (like stepper pulse pins, stepper direction pins, stepper enable pins) and are usually transversal to all CNC controlled machines.
- PWM pins - these are typical PWM with a resolution of 8bits (0-255)
- Servo pins -  these are also normal digital outputs that are controlled internally via software to generate a 50Hz pulses with high time between 0.5ms and 2.5ms. The resolution (range) is also 8bits (0-255)
- Generic digital output pins - these are all purpose pins that you can assign.

You can create you solenoid pins by assigning some of the generic output pins to your solenoids and then control them with a custom M380/M381 code using the generic io_set_pinvalue (for future version v1.8.x of µCNC) or if you are using version 1.7.x use function io_set_output.

You can instead use as an example the M42 module that controls outputs via M42 command,

All IO controlling functions are declared inside /src/core/io_control.h

**Comment by aquietude** [comment]
So, I’m facing one problem (kind of) and one doubt.
Your M42 module works fine for what I want but if I declare it inside load_modules function, like
`load_modules(
         LOAD_MODULE(m42);
);`
I get ‘too many arguments’ on function load_modules();
if I just delete load_modules(); and declare LOAD_MODULE(m42); it compiles fine though.

And my doubt is, I declared STEP0 to correspond to the physical port of A5 on my STM32 Blue Pill and I want that stepper to work as my X axis stepper, because it is connected to an X axis step motor, and my STEP1, corresponding to the physical port B0 of my board, connected to an Y axis step motor. How do I grant that if I run G1 X10 on UGS  it will move my STEP0 instead of somehow trying to misbehave STEP1?

**Comment by Paciente8159** [comment]
You are not declaring it correctly.

inside module.c do this

```
static FORCEINLINE void load_modules(void)
{
// PLACE YOUR MODULES HERE
#ifdef LOAD_MODULES_OVERRIDE
	LOAD_MODULES_OVERRIDE();
#endif

// add your module manually like this
LOAD_MODULE(m42);
}
```

About the stepper question I'm having a little trouble understanding the question but it's like this.

If you send G1 X10 you are sending a command to move axis X. Axis X will make the machine move in the X direction. If it pulses only STEP0, STEP0 and STEP1 or other combination of STEP outputs will depend on your kinematic and your configuration.

For example using the default configuration for a cartesian kinematic axis X will only make STEP0 move. Axis Y will only make STEP1 move and Axis Z will only make STEP2 move. That's it.

**Reply by aquietude** [reply]
That was exactly my doubt, thank you, I’m refactoring some Grbl_esp32 code to uCNC and on Grbl_esp32 that was set manually, like, for my A5 pin I would do X_STEP_PIN = GPIO5, thank you mate

**Comment by aquietude** [comment]
Hello again, I'm encountering an issue with serial monitoring/sending/recieving serial messages on my STM32 Blue Pill board running the uCNC firmware. I've configured and compiled the firmware using PlatformIO with the following settings:
```
[platformio]
include_dir=uCNC
src_dir=uCNC
default_envs = bluepill_f103c8
extra_configs = 
  uCNC/src/hal/boards/avr/avr.ini
  uCNC/src/hal/boards/samd21/samd21.ini
  uCNC/src/hal/boards/stm32/stm32.ini
  uCNC/src/hal/boards/esp8266/esp8266.ini
  uCNC/src/hal/boards/lpc176x/lpc176x.ini
  uCNC/src/hal/boards/esp32/esp32.ini
  uCNC/src/hal/boards/rp2040/rp2040.ini
;   uCNC/src/hal/mcus/virtual/virtual.ini

[common]
build_flags = -std=gnu99 -Wall -fdata-sections -ffunction-sections -fno-exceptions -Wl,--gc-sections

[env:bluepill_f103c8]
platform = ststm32
board = genericSTM32F103C8
upload_protocol = stlink
framework = arduino
build_flags = ${common.build_flags}
monitor_speed = 115200
monitor_eol = LF
monitor_echo = yes
monitor_filters = colorize, time
; uncomment to add graphic display library
; lib_deps = olikraus/U8g2
```
I'm using an Serial to USB conversor to do the serial monitoring. The issue is, my wirings, my conversor, and my STM32 are fine, and I know it because whenever I upload it with any piece of code, such as the following:
```
#include <Arduino.h>

   void setup() {
  Serial.begin(115200);
  while(!Serial);
}

void loop() {
  if (Serial.available()) {
    char data = Serial.read();
    
    Serial.write(data);
  }
  delay(0);
}
```
It works just fine, I can send serial messages and recieve them just fine for any of the RX/TX wirings of the STM32 BluePill, but whenever I try to send $I via serial with uCNC firmware, it just doesn't returns nothing, even if I set to send something and recieve something back if that something is sent via serial with Docklight (for example: if I send 1, recieve "1 recieved"). I tried it with PlatformIO Serial Monitor, Docklight and Arduino Serial Monitor. I even tried using the three of the RX/TX of the STM32 BluePill to do it , setting it using uCNC config generator and leaving the rest as defaults.

**Reply by Paciente8159** [reply]
If you configured the firmware with the default settings and you did not wire the control pins you are probably trying to use the firmware with the Emergency stop active.
Please [check if this is your case](https://github.com/Paciente8159/uCNC/wiki/FAQ#ok-ive-installed-%C2%B5cnc-on-my-board-but-nothing-is-happening-what-went-wrong)

**Comment by aquietude** [comment]
Hello there once more.
I did just what you said and wired the E-STOP pin to the GND of my board and it worked just fine, I was able to send ```$I```, ```$```, and some other system commands.
The problem is I'm still facing some issues which I have no idea why. I'll summarize those in topics to be easier for visualizing. 
- First of all, whenever I use my RX/TX alongside with an serial to USB conversor I just can't monitor or send anything by that COM, it opens, but it just doesn't send/recieve nothing, and I tested my RX/TX ports and my serial to USB conversor with other projects and also with an oscilloscope and I'm sure that it works. If I use an USB cable connected to the USB micro port of the STM32 BluePill and to a computer it just works fine, and I can assure that because I can send the $I command and it displays system info as excpeted. For testing this isn't a big deal but I'll ned to use the UART communication instead of USB for my CNC project so it would be good if I can somehow manage to fix it.
- Second, the G-code commands just does nothing. Whenever I send the G-code over console it parses as it worked just fine, and I know it because the uCNC returns me ```ok``` after the command but it doesn't do anything. My wiring to my machine is right and even if it wasn't I should be able to see the pulses of the command with the oscilloscope. To be more clear, the only commands I'm using are the ones related to moving the X-axis and the ones related to moving the Y-axis, such as ```G0``` and ```G1``` but whenever I send those it doesn't appear any pulse on the X-axis stepper pin neither on the Y-axis stepper pin and when wired to my machine it's stepper doesn't mover either.
- Third, some µCNC configurations commands just doesn't work for me, like ```$4=1``` and ```$4=0```, which should change the stepper enable from high to low and vice-versa but it just does plain nothing, the firmware returns parsing it as ```ok``` but the signal on the oscilloscope doesn't change. As well, the ```$22```, which should enable Homing cycle, which I was trying to enable in order to use Jog commands, but its more of the same, the firmware returns it as ```ok``` and Homing cycle just isn't enabled since I can't use Jog commands, whenever I try, Grbl returns that Homing cycle needs to be enabled in order to use jog commands. Also ```$H``` command, which is the Grbl command for enabling homing cycle just returns an error code.
- Fourth, I uncommented ```#define ENABLE_SETTING_EXTRA_CMDS``` on my ```cnc_config.h``` and recompilled the firmware on my board with it enabled and when I try to use the ```$P``` command it just returns an error like it wasn't enabled. 
- Fifth, whenever I try to use UGS as an G-code sender, when I connect to the COM it is mantained on ```HOLD``` and it displays ```ALARM X Y Z A PROBE HOLD``` and I'm unable to send any commands, neither via console neither via interface, when I try to use cncjs, perhaps, it connects just fine and I'm able to send commands via console, I'm just relating this because I think it can be of use regarding my problems.
- Sixth, I enabled the Marlin's ```M17``` and ```M18``` commands using you uCNC aditional modules in order to enable/disable the enable steppers pin, since the ```$4=1``` and ```$4=0``` didn't work, and it works just as excpeted and it appears on my oscilloscope the change of signal from high to low and vice-versa, I'm also relating it because I think it me of use somehow.

**Reply by Paciente8159** [reply]
Let's run the issues one by one:

> * First of all, whenever I use my RX/TX alongside with an serial to USB conversor I just can't monitor or send anything by that COM, it opens, but it just doesn't send/recieve nothing, and I tested my RX/TX ports and my serial to USB conversor with other projects and also with an oscilloscope and I'm sure that it works. If I use an USB cable connected to the USB micro port of the STM32 BluePill and to a computer it just works fine, and I can assure that because I can send the $I command and it displays system info as excpeted. For testing this isn't a big deal but I'll ned to use the UART communication instead of USB for my CNC project so it would be good if I can somehow manage to fix it.

This should work both in parallel. I will check this later today. On the default config both USB and COM1 (Pins A9 and A10) are enabled.
![image](https://github.com/Paciente8159/uCNC/assets/12116536/7803bfbe-6b18-4f72-86df-3d17336a7184)


> * Second, the G-code commands just does nothing. Whenever I send the G-code over console it parses as it worked just fine, and I know it because the uCNC returns me `ok` after the command but it doesn't do anything. My wiring to my machine is right and even if it wasn't I should be able to see the pulses of the command with the oscilloscope. To be more clear, the only commands I'm using are the ones related to moving the X-axis and the ones related to moving the Y-axis, such as `G0` and `G1` but whenever I send those it doesn't appear any pulse on the X-axis stepper pin neither on the Y-axis stepper pin and when wired to my machine it's stepper doesn't mover either.

At first I did not understand why this was happening but after reading the remaining issues I have a pretty good guess to what is going on. As stated in the FAQ page if not wired all limits and control pins will be active (usually on all default configurations these pins have an internal pull up enabled).
This means that also both Hold input and cycle start-resume will be enabled. Since Hold as a higher precedence the machine will be held on hold. You can confirm this by sending a `?` command to query for the machine status. If you inverted ESTOP you now need to invert the remaining control pins with setting [$7](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#7----control-pins-invert-mask). Just beware not to double invert ESTOP again.

> * Third, some µCNC configurations commands just doesn't work for me, like `$4=1` and `$4=0`, which should change the stepper enable from high to low and vice-versa but it just does plain nothing, the firmware returns parsing it as `ok` but the signal on the oscilloscope doesn't change. As well, the `$22`, which should enable Homing cycle, which I was trying to enable in order to use Jog commands, but its more of the same, the firmware returns it as `ok` and Homing cycle just isn't enabled since I can't use Jog commands, whenever I try, Grbl returns that Homing cycle needs to be enabled in order to use jog commands. Also `$H` command, which is the Grbl command for enabling homing cycle just returns an error code.

The machine does not JOG or HOME probably because it is in HOLD state. Also note that $4 is not a boolean but a bit mask similar to $3. I will correct that info in the wiki (one bit for each STEP EN pin). In this case this does not make a difference since STEP0_EN will control all steppers en pins. Just a few notes:

- If you enabled  ENABLE_SETTING_EXTRA_CMDS settings modifications will not go directly to FLASH until you save them with `$SS`.
- If you reflashed the machine previously, just to make sure there is not some incorrect setting somewhere do a `$RST=*` to reset the FLASH settings. If you do this and you had masked ESTOP via setting $7 you will need to re-do it again.

> * Fourth, I uncommented `#define ENABLE_SETTING_EXTRA_CMDS` on my `cnc_config.h` and recompilled the firmware on my board with it enabled and when I try to use the `$P` command it just returns an error like it wasn't enabled.

This should work. Again I think these issues are all related.

> * Fifth, whenever I try to use UGS as an G-code sender, when I connect to the COM it is mantained on `HOLD` and it displays `ALARM X Y Z A PROBE HOLD` and I'm unable to send any commands, neither via console neither via interface, when I try to use cncjs, perhaps, it connects just fine and I'm able to send commands via console, I'm just relating this because I think it can be of use regarding my problems.

I have no way of testing that right now but your machine is in ALARM mode. Has you can see also HOLD is active due to the unwired pin/unmasked setting.

Assuming that ESTOP is not active too. Try sending a `$X` command to unlock the machine. Again confirm the status with `?`

> * Sixth, I enabled the Marlin's `M17` and `M18` commands using you uCNC aditional modules in order to enable/disable the enable steppers pin, since the `$4=1` and `$4=0` didn't work, and it works just as excpeted and it appears on my oscilloscope the change of signal from high to low and vice-versa, I'm also relating it because I think it me of use somehow.

With the machine locked I don't think you will be able to do much. Check if solving the issues above also solves this issue.
